### PR TITLE
Makes Makefile template more modular.

### PR DESCRIPTION
**Pull request by halfak**
https://phabricator.wikimedia.org/T190968

### CODE REVIEW

File: editquality/codegen/generate.py [hunk 1]
```diff
@@ -1,5 +1,10 @@
 import jinja2
 
+ALG_FILE_NAME = {
```
Comment (adamwight) [inline editquality/codegen/generate.py]: This doesn't belong here, but what if we generalize a "globals.yaml" mechanism?

File: templates/Makefile.j2 [hunk 2]
```diff
@@ -21,116 +21,80 @@ include  Makefile.manual
 
 {% for wiki in wikis %}
 ############################# {{ wiki.label }} ################################
-{% for sample_name, sample in wiki.samples|dictsort %}
-
+{%if wiki.external_samples%}
```
Comment (adamwight) [inline templates/Makefile.j2]: I'm confused, this seems redundant with the "for" loop below, which will execute zero times for an empty parameter.  Was it throwing null errors or something?

(whitespaces inside the "%" for sanity)

File: templates/Makefile.j2 [hunk 3]
```diff
@@ -142,44 +106,33 @@ tuning_reports/{{ wiki.name }}.{{ model_name }}.md: \
...
+{% if model.cv_train %}
+models/{{ wiki.name }}.{{ model_name }}.{{ norm_alg_filename(model.cv_train.algorithm) }}.model: \
+		datasets/{{ wiki.name }}.
+{% endif %}.{{ wiki.name }}.{{ model_name }} \
 		{{ model.label }} \
 		--version=$({{ model_name }}_major_minor).{{ model.build_number }} \
-		{% for key, value in model.tuning_params|dictsort %}
-		-p '{{ key }}={{ value }}' \
+		{% for key, value in model.cv_train.parameters|dictsort %}
+		-p '{{ key }}={{ value|tojson|safe }}' \
```
Comment (adamwight) [inline templates/Makefile.j2]: Nice!

File: templates/Makefile.j2 [hunk 4]
```diff
@@ -142,44 +106,33 @@ tuning_reports/{{ wiki.name }}.{{ model_name }}.md: \
...
 		{% endfor %}
-		--label-weight "{{ model.label_weight }}" \
+		--label-weight $({{ model_name }}_label_weight) \
```
Comment (adamwight) [inline templates/Makefile.j2]: So this gets defined in the Makefile?  I don't see the declaration, though.

File: templates/Makefile.j2 [hunk 5]
```diff
@@ -142,44 +106,33 @@ tuning_reports/{{ wiki.name }}.{{ model_name }}.md: \
...
-		datasets/{{ wiki.name }}.{% if model_name == "reverted" %}auto{% endif %}labeled_revisions.w_cache.{{ wiki.default_sample }}.json
 {% endif %}
-	cat $^ | \
-	revscoring cv_train \
-		revscoring.scoring.models.{{ model.class_name }} \
-		editquality.feature_lists.{{ wiki.name }}.{{ model_name }} \
+{% if model.cv_train %}
+models/{{ wiki.name }}.{{ model_name }}.{{ norm_alg_filename(model.cv_train.algorithm) }}.model: \
+		datasets/{{ wiki.name }}.
+{% endif %}.{{ wiki.name }}.{{ model_name }} \
```
Comment (adamwight) [inline templates/Makefile.j2]: Something got jumbled here, looks like we're missing lines e.g. "revscoring cv_train"

File: templates/Makefile.j2 [hunk 6]
```diff
@@ -142,44 +106,33 @@ tuning_reports/{{ wiki.name }}.{{ model_name }}.md: \
...
+		{% for key, value in model.cv_train.parameters|dictsort %}
+		-p '{{ key }}={{ value|tojson|safe }}' \
 		{% endfor %}
-		--label-weight "{{ model.label_weight }}" \
+		--label-weight $({{ model_name }}_label_weight) \
 		--pop-rate "true={{ model.pop_rate_true }}" \
 		--pop-rate "false={{ 1 - model.pop_rate_true }}" \
-		--center --scale > $@
-	
+		--center --scale > $@;
```
Comment (adamwight) [inline templates/Makefile.j2]: Semicolon not required.

File: templates/Makefile.j2 [hunk 7]
```diff
@@ -142,44 +106,33 @@ tuning_reports/{{ wiki.name }}.{{ model_name }}.md: \
...
 
+{{ wiki.name }}_models:{% if wiki.models %} \
+{% for model_name, model in wiki.models.items() if model.cv_train %}
+	models/{{ wiki.name }}.{{ model_name }}.{{ norm_alg_filename(model.cv_train.algorithm) }}.model{% if not loop.last %} \{% endif %}
 {% endfor %}
 {% endif %}
 
 {{ wiki.name }}_tuning_reports:{% if wiki.models %} \
-{% for model_name, model in wiki.models.items() %}
+{% for model_name, model in wiki.models.items() if model.tune %}
```
Comment (adamwight) [inline templates/Makefile.j2]: Oh cool!

**Review by adamwight — DISMISSED** [review]
This is a great cleanup step, just a few more fixes remaining!

File: templates/Makefile.j2 (hunk 4 above)
Comment (halfak) [reply templates/Makefile.j2]: Yes.  Just like it currently is.

File: templates/Makefile.j2 (hunk 5 above)
Comment (halfak) [reply templates/Makefile.j2]: Definitely.  That doesn't make any sense.

File: editquality/codegen/generate.py (hunk 1 above)
Comment (Ladsgroup) [reply editquality/codegen/generate.py]: Agreed

Comment (halfak) [reply editquality/codegen/generate.py]: Hmm  What exactly are you proposing?  A `globals.yaml` where we still need to look up the `ALG_FILE_NAME` parameter?  What does that give us?

File: templates/Makefile.j2 (hunk 2 above)
Comment (halfak) [reply templates/Makefile.j2]: If we don't check, then we need to enforce a bunch of normalization in the config.py.  It gets messy, so I think it's just easier to check for existence here.

File: editquality/codegen/generate.py (hunk 1 above)
Comment (Ladsgroup) [reply editquality/codegen/generate.py]: It would separate configs from code. You can just inject value of the `ALG_FILE_NAME` in several ways.

Comment (halfak) [reply editquality/codegen/generate.py]: OK done.

**Comment by Ladsgroup** [comment]
I think this PR should stay a refactor and don't touch the output Makefile. The 2044 lines changed in the output is concerning to me and I don't want to merge something this big in one step.

**Comment by halfak** [comment]
We discussed this in IRC.  IMO, the Makefile is "internal code" and not really functionality.  The functionality is producing the outputs -- models, model_info, and tuning reports.  This merge is as big as reviewing the complete retraining of models -- which is something I reviewed in less than an hour just a couple weeks ago.

**Comment by halfak** [comment]
OK this is ready.  You can review changes to the models using this link: https://github.com/wikimedia/editquality/compare/modular_template...modular_template_info  There should be no substantial changes.

**Comment by halfak** [comment]
The only change I see is that we're now using 100k revisions to train nowiki.reverted rather than 40k revisions.  I'm not sure why we sampled down to 40k before.  We gain some fitness for training with 100k instead, so I'd like to keep it.  But in order to make this a pure refactor, we should probably drop that back to 40k.

**Comment by halfak** [comment]
Self merging per @Ladsgroup's approval (relayed by James Hare)

**Pull request merged**